From cb5faf15da683154a3847d9fb0bca47c51ff1f21 Mon Sep 17 00:00:00 2001
From: Richard Gong <richard.gong@intel.com>
Date: Tue, 22 Jun 2021 11:37:43 -0500
Subject: [PATCH 07/42] firmware: stratix10-svc: update for removing the
 service key object

commit 92c03561ee2cde1919e2ceede9455346f64826c3 from
https://github.com/altera-opensource/linux-socfpga/commits/socfpga-5.4.124-lts

Update to support removing the service key object.

Signed-off-by: Richard Gong <richard.gong@intel.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/firmware/stratix10-svc.c              |  7 ++++++
 include/linux/firmware/intel/stratix10-smc.h  | 24 +++++++++++++++++++
 .../firmware/intel/stratix10-svc-client.h     |  4 ++++
 3 files changed, 35 insertions(+)

diff --git a/drivers/firmware/stratix10-svc.c b/drivers/firmware/stratix10-svc.c
index 5ef8e244ba2a..3effac4ab2b9 100644
--- a/drivers/firmware/stratix10-svc.c
+++ b/drivers/firmware/stratix10-svc.c
@@ -334,6 +334,7 @@ static void svc_thread_recv_status_ok(struct stratix10_svc_data *p_data,
 	case COMMAND_FCS_ATTESTATION_CERTIFICATE_RELOAD:
 	case COMMAND_FCS_CRYPTO_CLOSE_SESSION:
 	case COMMAND_FCS_CRYPTO_IMPORT_KEY:
+	case COMMAND_FCS_CRYPTO_REMOVE_KEY:
 		cb_data->status = BIT(SVC_STATUS_OK);
 		break;
 	case COMMAND_RECONFIG_DATA_SUBMIT:
@@ -617,6 +618,11 @@ static int svc_normal_to_secure_thread(void *data)
 			a3 = (unsigned long)pdata->paddr_output;
 			a4 = (unsigned long)pdata->size_output;
 			break;
+		case COMMAND_FCS_CRYPTO_REMOVE_KEY:
+			a0 = INTEL_SIP_SMC_FCS_REMOVE_CRYPTO_SERVICE_KEY;
+			a1 = pdata->arg[0];
+			a2 = pdata->arg[1];
+			break;
 
 		/* for polling */
 		case COMMAND_POLL_SERVICE_STATUS:
@@ -707,6 +713,7 @@ static int svc_normal_to_secure_thread(void *data)
 			case COMMAND_FCS_CRYPTO_CLOSE_SESSION:
 			case COMMAND_FCS_CRYPTO_IMPORT_KEY:
 			case COMMAND_FCS_CRYPTO_EXPORT_KEY:
+			case COMMAND_FCS_CRYPTO_REMOVE_KEY:
 				cbdata->status = BIT(SVC_STATUS_INVALID_PARAM);
 				cbdata->kaddr1 = NULL;
 				cbdata->kaddr2 = NULL;
diff --git a/include/linux/firmware/intel/stratix10-smc.h b/include/linux/firmware/intel/stratix10-smc.h
index 029288e8fb6d..61d6f957fd19 100644
--- a/include/linux/firmware/intel/stratix10-smc.h
+++ b/include/linux/firmware/intel/stratix10-smc.h
@@ -872,4 +872,28 @@ INTEL_SIP_SMC_FAST_CALL_VAL(INTEL_SIP_SMC_FUNCID_FPGA_CONFIG_COMPLETED_WRITE)
 #define INTEL_SIP_SMC_FCS_EXPORT_CRYPTO_SERVICE_KEY \
         INTEL_SIP_SMC_FAST_CALL_VAL(INTEL_SIP_SMC_FUNCID_FCS_EXPORT_CRYPTO_SERVICE_KEY)
 
+/**
+ * Request INTEL_SIP_SMC_FCS_REMOVE_CRYPTO_SERVICE_KEY
+ * Sync call to remove the crypto service kers from the device
+ *
+ * Call register usage:
+ * a0 INTEL_SIP_SMC_FCS_REMOVE_CRYPTO_SERVICE_KEY
+ * a1 session ID
+ * a2 key UID
+ * a3-a7 not used
+ *
+ * Return status:
+ * a0 INTEL_SIP_SMC_STATUS_OK, INTEL_SIP_SMC_STATUS_NOT_SUPPORTED or
+ *    INTEL_SIP_SMC_STATUS_ERROR
+ * a1 mailbox and status errors if a0 is INTEL_SIP_SMC_STATUS_ERROR
+ * 	31:24 -- reserved
+ * 	23:16 -- import/export/removal status error
+ * 	15:11 -- reserved
+ * 	10:0  -- mailbox error
+ * a2-a3 not used
+ */
+#define INTEL_SIP_SMC_FUNCID_FCS_REMOVE_CRYPTO_SERVICE_KEY 114
+#define INTEL_SIP_SMC_FCS_REMOVE_CRYPTO_SERVICE_KEY \
+	INTEL_SIP_SMC_FAST_CALL_VAL(INTEL_SIP_SMC_FUNCID_FCS_REMOVE_CRYPTO_SERVICE_KEY)
+
 #endif
diff --git a/include/linux/firmware/intel/stratix10-svc-client.h b/include/linux/firmware/intel/stratix10-svc-client.h
index b6b4446d9d84..f456c0211fe7 100644
--- a/include/linux/firmware/intel/stratix10-svc-client.h
+++ b/include/linux/firmware/intel/stratix10-svc-client.h
@@ -180,6 +180,9 @@ struct stratix10_svc_chan;
  *
  * @COMMAND_FCS_CRYPTO_EXPORT_KEY: export the crypto service key object,
  * return status is SVC_STATUS_OK or SVC_STATUS_ERROR
+ *
+ * @COMMAND_FCS_CRYPTO_REMOVE_KEY: remove the crypto service key object
+ * from the device, return status is SVC_STATUS_OK or SVC_STATUS_ERROR
  */
 enum stratix10_svc_command_code {
 	/* for FPGA */
@@ -223,6 +226,7 @@ enum stratix10_svc_command_code {
 	COMMAND_FCS_CRYPTO_CLOSE_SESSION,
 	COMMAND_FCS_CRYPTO_IMPORT_KEY,
 	COMMAND_FCS_CRYPTO_EXPORT_KEY,
+	COMMAND_FCS_CRYPTO_REMOVE_KEY,
 };
 
 /**
-- 
2.31.1

